plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,155 @@
+/*
+*	Copyright (C) 2017-2017 Alessandro Vetere
+*	Copyright (C) 2007-2009 Gabest
+*	http://www.gabest.org
+*
+*  This Program is free software; you can redistribute it and/or modify
+*  it under the terms of the GNU General Public License as published by
+*  the Free Software Foundation; either version 2, or (at your option)
+*  any later version.
+*
+*  This Program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+*  GNU General Public License for more details.
+*
+*  You should have received a copy of the GNU General Public License
+*  along with GNU Make; see the file COPYING.  If not, write to
+*  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA USA.
+*  http://www.gnu.org/copyleft/gpl.html
+*
+*/
+
+#pragma once
+
+template <class T>
+struct Element {
+	T  data;
+	uint16 next_index;
+	uint16 prev_index;
+};
+
+// Class size in bytes:
+// 2 * sizeof(uint16) + sizeof(Element<T>*) + sizeof(uint16*),
+// The Element<T> array m_buffer has dynamic size: CURRENT_SIZE,
+// The uint16 array m_free_indexes_stack has dynamic size: CURRENT_SIZE - 1,
+// Where CURRENT_SIZE starts from INITIAL_SIZE and grows as needed
+template <class T, uint16 INITAL_SIZE = 4>
+class FastList {
+public:
+	// The last Element<T> of the list has next_index equal to END
+	uint16 END;
+	// m_buffer[0] is the head of the list, and m_buffer has END elements
+	Element<T>* m_buffer;
+
+	FastList() {
+		m_buffer = NULL;
+		Clear();
+	}
+
+	~FastList() {
+		_aligned_free(m_buffer);
+	}
+
+	void Clear() {
+		// Initialize END
+		END = INITAL_SIZE;
+		// Initialize arrays
+		m_buffer = (Element<T>*)_aligned_realloc(m_buffer, END * sizeof(Element<T>) + (END - 1) * sizeof(uint16), 64);
+		m_free_indexes_stack = (uint16*)&m_buffer[END];
+		// Initialize head
+		m_buffer[0].data = NULL;
+		m_buffer[0].next_index = END;
+		m_buffer[0].prev_index = 65535;
+		// Stack top is 0, bottom is END - 2
+		m_free_indexes_stack_top = 0;
+		// Index 0 reserved for head
+		for (uint16 i = 0; i < END - 1; i++) {
+			m_free_indexes_stack[i] = i + 1;
+		}
+	}
+
+	__forceinline bool Empty() {
+		return m_free_indexes_stack == 0;
```

Comment:
Obviously it's `m_free_indexes_stack_top` and not `m_free_indexes_stack` that should be 0.